[iOS] Add picker UI for <input type=week>

lilyspiniolas opened the pull request:
#### ec3e75aa2e70528efbc7e0e9f7c4f8ad9b2ccb44
<pre>
[iOS] Add picker UI for &lt;input type=week&gt;
<a href="https://bugs.webkit.org/show_bug.cgi?id=241598">https://bugs.webkit.org/show_bug.cgi?id=241598</a>
<a href="https://rdar.apple.com/10854201">rdar://10854201</a>

Reviewed by Aditya Keerthi.

Clicking the week input on iOS now displays a UICalendarView with a week selection type rather than simply functioning as a text field.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/html/DateTimeFieldsState.h:
* Source/WebCore/platform/text/cocoa/LocaleCocoa.mm:
(WebCore::LocaleCocoa::formatDateTime):
* Source/WebCore/platform/text/cocoa/LocalizedDateCache.mm:
(WebCore::LocalizedDateCache::createFormatterForType):
(WebCore::LocalizedDateCache::calculateMaximumWidth):
* Source/WebCore/rendering/ios/RenderThemeIOS.mm:
(WebCore::adjustInputElementButtonStyle):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _shouldShowAutomaticKeyboardUIIgnoringInputMode]):
(-[WKContentView _elementTypeRequiresAccessoryView:]):
(mayContainSelectableText):
(createInputPeripheralWithView):
(-[WKContentView _createTargetedContextMenuHintPreviewForFocusedElement:]):
* Source/WebKit/UIProcess/ios/forms/WKDatePickerPopoverController.h:
* Source/WebKit/UIProcess/ios/forms/WKDatePickerPopoverController.mm:
(-[WKDatePickerPopoverView setupView:toolbarBottomMargin:]):
(-[WKDatePickerPopoverView initWithCalendarView:selectionWeekOfYear:]):
(-[WKDatePickerPopoverView initWithDatePicker:]):
(-[WKDatePickerPopoverView calendarView]):
(-[WKDatePickerPopoverController initWithCalendarView:selectionWeekOfYear:delegate:]):
* Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
(-[WKDateTimePicker initWithView:inputType:]):
(-[WKDateTimePicker weekOfYearSelection:didSelectWeekOfYear:]):
(-[WKDateTimePicker showDateTimePicker]):
(-[WKDateTimePicker iso8601DateFormatterForCalendarView]):
(-[WKDateTimePicker _dateChanged]):
(-[WKDateTimePicker setWeekPickerToInitialValue]):
(-[WKDateTimePicker setDateTimePickerToInitialValue]):
(-[WKDateTimeInputControl initWithView:]):
(-[WKDateTimePicker initWithView:datePickerMode:]): Deleted.
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::isAssistableElement):

Canonical link: <a href="https://commits.webkit.org/281437@main">https://commits.webkit.org/281437@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0faad36c890a6febd0a56ed6d3a87d930aecf57a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59904 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/39252 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/12483 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/63822 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/10432 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/62033 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/46935 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/10661 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/48555 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7280 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61934 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/36623 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/51895 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/29399 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33329 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/9135 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/9353 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/52998 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/55245 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9417 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/65553 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/59149 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/3834 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/9280 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/55901 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/3879 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51887 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56045 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13278 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3201 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/80907 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/35065 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14047 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/36147 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/37235 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/35891 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebCore/html/DateTimeFieldsState.h:
```diff
@@ -43,6 +43,7 @@ struct DateTimeFieldsState {
     }
 
     std::optional<unsigned> year;
+    std::optional<unsigned> week;
```
this is the most nit-picky thing ever, but I think this should be under month, since the order of these seems to be largest unit first

rr-codes commented on Source/WebKit/UIProcess/ios/forms/WKDatePickerPopoverController.mm:
```diff
@@ -96,16 +93,43 @@ - (instancetype)initWithDatePicker:(UIDatePicker *)datePicker
+        return nil;
+
+    _calendarView = calendarView;
+    NSCalendar *cal = [NSCalendar calendarWithIdentifier:NSCalendarIdentifierISO8601];
```
we usually try to avoid abbreviations; also, this should be `RetainPtr name` instead of `NSCalendar *name`

aprotyas replied:
Is that a +1 method?

whsieh replied:
It's not a +1 method. Smart pointer guidelines would recommend using RetainPtr here though (`RetainPtr calendar = [NSCalendar calendarWithIdentifier:NSCalendarIdentifierISO8601];`)

rr-codes commented on Source/WebKit/UIProcess/ios/forms/WKDatePickerPopoverController.mm:
```diff
@@ -96,16 +93,43 @@ - (instancetype)initWithDatePicker:(UIDatePicker *)datePicker
+
+    _calendarView = calendarView;
+    NSCalendar *cal = [NSCalendar calendarWithIdentifier:NSCalendarIdentifierISO8601];
+    [cal setFirstWeekday:2];
```
should this be hard-coded like this?

rr-codes commented on Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
```diff
@@ -192,6 +228,14 @@ - (NSString *)_sanitizeInputValueForFormatter:(NSString *)value
 
+- (RetainPtr<NSISO8601DateFormatter>)iso8601DateFormatterForCalendarView
+{
+    auto dateFormatter = adoptNS([[NSISO8601DateFormatter alloc] init]);
```
`RetainPtr` instead of `auto`

rr-codes commented on Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
```diff
@@ -192,6 +228,14 @@ - (NSString *)_sanitizeInputValueForFormatter:(NSString *)value
+{
+    auto dateFormatter = adoptNS([[NSISO8601DateFormatter alloc] init]);
+    [dateFormatter setTimeZone:[_calendarView timeZone]];
+    dateFormatter.get().formatOptions = NSISO8601DateFormatWithYear | NSISO8601DateFormatWithWeekOfYear | NSISO8601DateFormatWithDashSeparatorInDate;
```
I think we usually do `[dateFormatter setFormatOptions:` instead of `.get().formatOptions`, but it doesn't really matter

pxlcoder replied:
Yes, please use bracket syntax here.

rr-codes commented on Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
```diff
@@ -205,12 +249,38 @@ - (NSString *)_sanitizeInputValueForFormatter:(NSString *)value
 
+- (void)setWeekPickerToInitialValue
+{
+    unsigned unitFlags = NSCalendarUnitYearForWeekOfYear | NSCalendarUnitWeekOfYear;
```
I think this can be `NSCalendarUnit` instead of `unsigned`

rr-codes commented on Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
```diff
@@ -205,12 +249,38 @@ - (NSString *)_sanitizeInputValueForFormatter:(NSString *)value
+        return;
+    }
+
+    RetainPtr<NSDate> parsedDate = [[self iso8601DateFormatterForCalendarView] dateFromString:[self _sanitizeInputValueForFormatter:_initialValue.get()]];
```
I think the template `<NSData>` parameter can be removed

aprotyas commented on Source/WebCore/platform/text/cocoa/LocalizedDateCache.mm:
```diff
@@ -122,7 +122,8 @@ static void _localeChanged(CFNotificationCenterRef, void*, CFStringRef, const vo
         break;
     case DateComponentsType::Week:
-        ASSERT_NOT_REACHED();
+        [dateFormatter setTimeZone:[NSTimeZone timeZoneForSecondsFromGMT:0]];
```
I know you're just following the pattern here, but it looks like we make this same call for all cases... maybe pull it out?

aprotyas commented on Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
```diff
@@ -90,6 +95,10 @@ - (id)initWithView:(WKContentView *)view datePickerMode:(UIDatePickerMode)mode
         _formatString = kDateTimeFormatString;
         break;
+    case WebKit::InputType::Week:
+        ASSERT_NOT_REACHED();
```
Why the assert?

aprotyas reviewed (commented):
+1 to all of Richard's comments. This looks good! Though this isn't a comprehensive review

whsieh commented on Source/WebKit/UIProcess/ios/forms/WKDatePickerPopoverController.mm:
```diff
@@ -96,16 +93,43 @@ - (instancetype)initWithDatePicker:(UIDatePicker *)datePicker
+    NSCalendar *cal = [NSCalendar calendarWithIdentifier:NSCalendarIdentifierISO8601];
+    [cal setFirstWeekday:2];
+    [_calendarView setCalendar:cal];
+    [_calendarView setSelectionBehavior: weekSelection];
```
Nit - extra space before `weekSelection`

whsieh commented on Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
```diff
@@ -113,6 +122,30 @@ - (id)initWithView:(WKContentView *)view datePickerMode:(UIDatePickerMode)mode
     return self;
 }
 
+- (id)initWithView:(WKContentView *)view
```
Nit - `instancetype` as the return type

whsieh commented on Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
```diff
@@ -113,6 +122,30 @@ - (id)initWithView:(WKContentView *)view datePickerMode:(UIDatePickerMode)mode
+    _interactionPoint = [_view lastInteractionLocation];
+    _formatString = kWeekFormatString;
+
+    _selectionWeekOfYear = [[UICalendarSelectionWeekOfYear alloc] initWithDelegate:self];
```
This leaks:
```suggestion
    _selectionWeekOfYear = adoptNS([[UICalendarSelectionWeekOfYear alloc] initWithDelegate:self]);
```

whsieh commented on Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
```diff
@@ -113,6 +122,30 @@ - (id)initWithView:(WKContentView *)view datePickerMode:(UIDatePickerMode)mode
+    return self;
+}
+
+- (void)weekOfYearSelection:(UICalendarSelectionWeekOfYear *)selection didSelectWeekOfYear:(NSDateComponents*)weekOfYearComponents {
```
Minor whitespace nit -

```suggestion
- (void)weekOfYearSelection:(UICalendarSelectionWeekOfYear *)selection didSelectWeekOfYear:(NSDateComponents *)weekOfYearComponents {
```

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
```diff
@@ -35,23 +35,27 @@
 }
 
 - (instancetype)initWithView:(WKContentView *)view datePickerMode:(UIDatePickerMode)mode;
+- (instancetype)initWithView:(WKContentView *)view;
```
Why do we need a separate initializer?

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKDatePickerPopoverController.mm:
```diff
@@ -155,6 +184,18 @@ - (instancetype)initWithDatePicker:(UIDatePicker *)datePicker delegate:(id<WKDat
     return self;
 }
 
+- (instancetype)initWithCalendarView:(UICalendarView*)calendarView selectionWeekOfYear:(UICalendarSelectionWeekOfYear*)weekSelection delegate:(id<WKDatePickerPopoverControllerDelegate>)delegate
```
```suggestion
- (instancetype)initWithCalendarView:(UICalendarView*)calendarView selectionWeekOfYear:(UICalendarSelectionWeekOfYear *)weekSelection delegate:(id<WKDatePickerPopoverControllerDelegate>)delegate
```

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKDatePickerPopoverController.mm:
```diff
@@ -38,38 +38,35 @@
 @interface WKDatePickerPopoverView : UIView
 
 @property (readonly, nonatomic) UIDatePicker *datePicker;
+@property (readonly, nonatomic) UIDatePicker *calendarView;
```
I don't think this declaration is necessary? From a quick scan, I didn't see any code outside this class which accesses `calendarView`.

(also, the type is wrong)

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKDatePickerPopoverController.mm:
```diff
@@ -96,16 +93,43 @@ - (instancetype)initWithDatePicker:(UIDatePicker *)datePicker
+    [_calendarView setCalendar:cal];
+    [_calendarView setSelectionBehavior: weekSelection];
+
+    CGFloat toolbarBottomMargin = 2;
```
Probably time for a constant at the top of this file, since it's used in two different places.

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
```diff
@@ -113,6 +122,30 @@ - (id)initWithView:(WKContentView *)view datePickerMode:(UIDatePickerMode)mode
+
+    _selectionWeekOfYear = [[UICalendarSelectionWeekOfYear alloc] initWithDelegate:self];
+
+    _calendarView = adoptNS([[UICalendarView alloc]init]);
```
```suggestion
    _calendarView = adoptNS([[UICalendarView alloc] init]);
```

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
```diff
@@ -309,6 +379,9 @@ - (instancetype)initWithView:(WKContentView *)view
         mode = yearAndMonthDatePickerMode;
         break;
+    case WebKit::InputType::Week:
+        self = [super initWithView:view control:adoptNS([[WKDateTimePicker alloc] initWithView:view])];
```
Echoing my comment from above, I don't think we need a separate initializer here.

pxlcoder reviewed (commented):
I would change the commit title to:

> [iOS] Add picker UI for `<input type=week>`

(implementation details should be left for the commit message)

pxlcoder commented on Source/WTF/wtf/PlatformHave.h:
```diff
@@ -1853,3 +1853,9 @@
+#if !defined(HAVE_UICALENDARSELECTIONWEEKOFYEAR) \
+    && (((PLATFORM(IOS) || PLATFORM(MACCATALYST)) && __IPHONE_OS_VERSION_MIN_REQUIRED >= 180000) \
+    || (PLATFORM(VISION) && __VISION_OS_VERSION_MIN_REQUIRED >= 20000))
+#define HAVE_UICALENDARSELECTIONWEEKOFYEAR 1
```
Nit: Should probably be `HAVE_UICALENDAR_SELECTION_WEEK_OF_YEAR`.

pxlcoder commented on Source/WebCore/platform/text/cocoa/LocaleCocoa.mm:
```diff
@@ -89,11 +89,12 @@
 
-    // "week" type not supported.
     ASSERT(type != DateComponentsType::Invalid);
+#if !HAVE(UICALENDARSELECTIONWEEKOFYEAR)
```
I don't really like the idea of referencing a UIKit `HAVE` down in WebCore code that is UI-independent.

Could you also add an `ENABLE_INPUT_TYPE_WEEK_PICKER` flag for use in WebCore and in WebKit outside of the UI code? Eventually other platforms could also flip that on.

pxlcoder commented on Source/WebCore/platform/text/cocoa/LocalizedDateCache.mm:
```diff
@@ -155,10 +155,15 @@ static void _localeChanged(CFNotificationCenterRef, void*, CFStringRef, const vo
     ASSERT(numberOfGregorianMonths == 12);
 
     // For each month (in the Gregorian Calendar), format a date and measure its length.
+    // FIXME: Week is checked here as a workaround for an issue which caused the displayed text to overflow onto a second line.
```
I don't think it's a workaround? From my understanding, it's essential and something we intend to keep around.

lilyspiniolas replied:
Isn't that code meant to calculate the maximum width based on month names, though? I thought it was a workaround since the month names aren't displayed on the week picker's text (i.e "Week 29 2024")

pxlcoder replied:
It's not specifically about whether the month is displayed or not, it's just whether the month can influence the width of the displayed string which is true for weeks.

It could probably be further "optimized", since we know the maximum width would be for months with two-digit weeks. But this is not particularly hot code and I don't think it's something we need to come back to once this patch has landed.

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKDatePickerPopoverController.h:
```diff
@@ -38,6 +38,9 @@
 @interface WKDatePickerPopoverController : UIViewController
 - (instancetype)initWithDatePicker:(UIDatePicker *)datePicker delegate:(id<WKDatePickerPopoverControllerDelegate>)delegate;
+#if HAVE(UICALENDARSELECTIONWEEKOFYEAR)
+- (instancetype)initWithCalendarView:(UICalendarView *)calendarView selectionWeekOfYear:(UICalendarSelectionWeekOfYear*)weekSelection delegate:(id<WKDatePickerPopoverControllerDelegate>)delegate;
```
```suggestion
- (instancetype)initWithCalendarView:(UICalendarView *)calendarView selectionWeekOfYear:(UICalendarSelectionWeekOfYear *)weekSelection delegate:(id<WKDatePickerPopoverControllerDelegate>)delegate;
```

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKDatePickerPopoverController.mm:
```diff
@@ -45,31 +45,29 @@ @interface WKDatePickerPopoverView : UIView
 }
 
-- (instancetype)initWithDatePicker:(UIDatePicker *)datePicker
+- (void)setupView:(UIView *)mainView toolbarBottomMargin:(CGFloat)toolbarBottomMargin
```
```suggestion
- (void)setupView:(UIView *)pickerView toolbarBottomMargin:(CGFloat)toolbarBottomMargin
```

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKDatePickerPopoverController.mm:
```diff
@@ -96,16 +93,42 @@ - (instancetype)initWithDatePicker:(UIDatePicker *)datePicker
     ]];
+}
+
+#if HAVE(UICALENDARSELECTIONWEEKOFYEAR)
```
```suggestion
#if HAVE(UICALENDARSELECTIONWEEKOFYEAR)

```

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKDatePickerPopoverController.mm:
```diff
@@ -96,16 +93,42 @@ - (instancetype)initWithDatePicker:(UIDatePicker *)datePicker
+    [_calendarView setCalendar:[NSCalendar calendarWithIdentifier:NSCalendarIdentifierISO8601]];
+    [_calendarView setSelectionBehavior:weekSelection];
+
+    [self setupView:calendarView toolbarBottomMargin:2];
```
I think I had left an earlier comment about pulling out the `2` into a constant at the top of the file, since it's used in two places.

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
```diff
@@ -67,31 +77,47 @@ @implementation WKDateTimePicker
 
-    switch (view.focusedElementInformation.elementType) {
+    switch (inputType) {
     case WebKit::InputType::Date:
+        mode = UIDatePickerModeDate;
         _formatString = kDateFormatString;
         break;
     case WebKit::InputType::Month:
+        mode = yearAndMonthDatePickerMode;
         _formatString = kMonthFormatString;
         break;
     case WebKit::InputType::Time:
+        mode = UIDatePickerModeTime;
         _formatString = kTimeFormatString;
         break;
     case WebKit::InputType::DateTimeLocal:
+        mode = UIDatePickerModeDateAndTime;
         _formatString = kDateTimeFormatString;
         break;
+#if HAVE(UICALENDARSELECTIONWEEKOFYEAR)
+    case WebKit::InputType::Week:
+        _formatString = kWeekFormatString;
+        _selectionWeekOfYear = adoptNS([[UICalendarSelectionWeekOfYear alloc] initWithDelegate:self]);
+        _calendarView = adoptNS([[UICalendarView alloc] init]);
+        [_calendarView setCalendar:[NSCalendar calendarWithIdentifier:NSCalendarIdentifierISO8601]];
+        [_calendarView setSelectionBehavior:_selectionWeekOfYear.get()];
+        return self;
+#endif
     default:
-        break;
+        [self release];
+        return nil;
```
Is this change necessary?

lilyspiniolas replied:
It was changed because otherwise `mode` would be used uninitialized since at that point it didn't match any of the targeted input types. Would it be better to set it some value anyways?

pxlcoder replied:
I see! I think it's fine then, we don't expect this code to run anyway.

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKDatePickerPopoverController.mm:
```diff
@@ -96,16 +93,42 @@ - (instancetype)initWithDatePicker:(UIDatePicker *)datePicker
+
+    return self;
+}
+#endif
```
```suggestion

#endif
```

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
```diff
@@ -205,12 +252,44 @@ - (NSString *)_sanitizeInputValueForFormatter:(NSString *)value
+        [_view updateFocusedElementValue:[dateFormatter stringFromDate:[[NSCalendar calendarWithIdentifier:NSCalendarIdentifierISO8601] dateFromComponents:[_selectionWeekOfYear selectedWeekOfYear]]]];
+        return;
+    }
+#endif
```
```suggestion
#endif

```

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
```diff
@@ -205,12 +252,44 @@ - (NSString *)_sanitizeInputValueForFormatter:(NSString *)value
+    RetainPtr<NSDateComponents> dateComponents = [[NSCalendar calendarWithIdentifier:NSCalendarIdentifierISO8601] components:unitFlags fromDate:parsedDate.get() ? parsedDate.get() : [NSDate date]];
+
+    // FIXME: The selectedWeekOfYear from these components causes the week before the correct one to be selected, so it is manually adjusted here. There is likely a better way of achieving this.
+    dateComponents.get().weekOfYear += 1;
```
This seems a little concerning. Have you tested in device locales where the first weekday differs/matches from ISO8601?

In these cases we want to make sure what the user sees (both in-page, and the picker) matches the week number in their locale. This number could be different from what is exposed via JavaScript / HTML value.

pxlcoder commented on Source/WebCore/platform/text/cocoa/LocalizedDateCache.mm:
```diff
@@ -97,32 +97,32 @@ static void _localeChanged(CFNotificationCenterRef, void*, CFStringRef, const vo
+#if !HAVE(UI_CALENDAR_SELECTION_WEEK_OF_YEAR)
         ASSERT_NOT_REACHED();
+#endif
+        [dateFormatter setDateFormat:@"'Week' w',' yyyy"];
```
We still need localization here.

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
```diff
@@ -192,25 +233,69 @@ - (NSString *)_sanitizeInputValueForFormatter:(NSString *)value
     [dateFormatter setDateFormat:_formatString];
+    // Necessary in order to append monday to the initial value for week inputs so that the NSDate reflects weeks beginning on Mondays in ISO-8601
+    if (_view.focusedElementInformation.elementType == WebKit::InputType::Week)
+        [dateFormatter setDateFormat:@"yyyy-'W'ww-EEEEE"];
```
Why not just set `_formatString` to this value, in the initializer?

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
```diff
@@ -192,25 +233,69 @@ - (NSString *)_sanitizeInputValueForFormatter:(NSString *)value
+        return;
+    }
+
+    // TODO: There may be a better way of ensuring the calculated NSDate from the initial value reflects ISO-8601 weeks starting on Monday.
```
If we're going to merge this patch with this logic, this should be a `FIXME:` rather than a `TODO:`.

https://webkit.org/code-style-guidelines/#comments-fixme

pxlcoder reviewed (commented):
Commit message still needs to be updated. Are the other TODOs mentioned in the message complete?

pxlcoder commented on Source/WebCore/platform/text/cocoa/LocalizedDateCache.mm:
```diff
@@ -97,33 +98,37 @@ static void _localeChanged(CFNotificationCenterRef, void*, CFStringRef, const vo
     case DateComponentsType::Week:
+#if HAVE(UI_CALENDAR_SELECTION_WEEK_OF_YEAR)
+        RetainPtr weekFormatString = [NSMutableString stringWithString:@"'' w',' yyyy"];
+        [weekFormatString.get() insertString:WebCore::inputWeekLabel() atIndex:1];
```
`-[NSString stringWithFormat:]` would be better than creating a mutable string.

pxlcoder replied:
`[NSString stringWithFormat:@"'%@' w',' yyyy", inputWeekLabel()]`

pxlcoder replied:
The `WebCore::` is unnecessary since we're in the namespace.

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -2654,6 +2654,9 @@ - (BOOL)_shouldShowAutomaticKeyboardUIIgnoringInputMode
     case WebKit::InputType::Month:
     case WebKit::InputType::DateTimeLocal:
     case WebKit::InputType::Time:
+#if HAVE(UI_CALENDAR_SELECTION_WEEK_OF_YEAR)
```
I'd use the `ENABLE` in this file, since we're not directly using the API.

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
```diff
@@ -67,31 +77,48 @@ @implementation WKDateTimePicker
 static NSString * const kMonthFormatString = @"yyyy-MM"; // "2011-01".
 static NSString * const kTimeFormatString = @"HH:mm"; // "13:45".
 static NSString * const kDateTimeFormatString = @"yyyy-MM-dd'T'HH:mm"; // "2011-01-27T13:45"
+// Weekday is necessary for week inputs in order to append monday to the initial value so that the NSDate reflects weeks beginning on Mondays in ISO-8601
```
```suggestion
// Weekday is necessary for week inputs in order to append Monday to the initial value so that the NSDate reflects weeks beginning on Mondays in ISO-8601.
```

pxlcoder commented on Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
```diff
@@ -67,31 +77,48 @@ @implementation WKDateTimePicker
     _view = view;
     _interactionPoint = [_view lastInteractionLocation];
+    UIDatePickerMode mode;
 
```
```suggestion

    UIDatePickerMode mode;
```

pxlcoder reviewed (commented):
I think this is almost ready!

lilyspiniolas committed:
[iOS] Add picker UI for <input type=week>
https://bugs.webkit.org/show_bug.cgi?id=241598
rdar://10854201

Reviewed by Aditya Keerthi.

Clicking the week input on iOS now displays a UICalendarView with a week selection type rather than simply functioning as a text field.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/html/DateTimeFieldsState.h:
* Source/WebCore/platform/text/cocoa/LocaleCocoa.mm:
(WebCore::LocaleCocoa::formatDateTime):
* Source/WebCore/platform/text/cocoa/LocalizedDateCache.mm:
(WebCore::LocalizedDateCache::createFormatterForType):
(WebCore::LocalizedDateCache::calculateMaximumWidth):
* Source/WebCore/rendering/ios/RenderThemeIOS.mm:
(WebCore::adjustInputElementButtonStyle):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _shouldShowAutomaticKeyboardUIIgnoringInputMode]):
(-[WKContentView _elementTypeRequiresAccessoryView:]):
(mayContainSelectableText):
(createInputPeripheralWithView):
(-[WKContentView _createTargetedContextMenuHintPreviewForFocusedElement:]):
* Source/WebKit/UIProcess/ios/forms/WKDatePickerPopoverController.h:
* Source/WebKit/UIProcess/ios/forms/WKDatePickerPopoverController.mm:
(-[WKDatePickerPopoverView setupView:toolbarBottomMargin:]):
(-[WKDatePickerPopoverView initWithCalendarView:selectionWeekOfYear:]):
(-[WKDatePickerPopoverView initWithDatePicker:]):
(-[WKDatePickerPopoverView calendarView]):
(-[WKDatePickerPopoverController initWithCalendarView:selectionWeekOfYear:delegate:]):
* Source/WebKit/UIProcess/ios/forms/WKDateTimeInputControl.mm:
(-[WKDateTimePicker initWithView:inputType:]):
(-[WKDateTimePicker weekOfYearSelection:didSelectWeekOfYear:]):
(-[WKDateTimePicker showDateTimePicker]):
(-[WKDateTimePicker iso8601DateFormatterForCalendarView]):
(-[WKDateTimePicker _dateChanged]):
(-[WKDateTimePicker setWeekPickerToInitialValue]):
(-[WKDateTimePicker setDateTimePickerToInitialValue]):
(-[WKDateTimeInputControl initWithView:]):
(-[WKDateTimePicker initWithView:datePickerMode:]): Deleted.
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::isAssistableElement):

Canonical link: https://commits.webkit.org/281437@main
